engine/security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/BaseOAuth2Service.scala

```diff
@@ -0,0 +1,56 @@
+package pl.touk.nussknacker.ui.security.oauth2
+
+import com.typesafe.scalalogging.LazyLogging
+import io.circe.Decoder
+import pl.touk.nussknacker.engine.util.cache.{CacheConfig, DefaultAsyncCache, ExpiryConfig}
+import pl.touk.nussknacker.ui.security.api.LoggedUser
+
+import scala.concurrent.duration.{Deadline, FiniteDuration, MILLISECONDS}
+import scala.concurrent.{ExecutionContext, Future}
+
+class BaseOAuth2Service[
```

Comment:
Do you think there is a case for a service without even a short cache? With no cache, every request to the Nussknacker API triggers a query to a remote authorization server, which might be remote or have a quota.

Reply:
1. I can imagine that case.. But you're right it's unlikely.But still it's good practice to separate getting from cache.
2. I think this class should be named `CacheOAuth2Service` with possibility to inject `authorizationsCache` by constructor. Right now you support only one cache - `DefaultAsyncCache`. What about use db as a cache? Cache in memory will not work with more nk instances.

Reply:
You're right about most of the above.

Reply:
I tried to convert it your way, but it is easier said than done.
With the decorator pattern, cache entries would have to be created from the results of methods of the decorated object only.
The first problem is with token expiry which in the case of JWT would need to be returned from the `authorize` method, this would be justified.
But another problem is with OIDC. It is nice to use an id token that comes with the access token response and not to call the `/userinfo` endpoint later. Again, to cache with a decorator, we would need to have the user profile returned from the service's method along with an expiry. In my opinion, this would make the pattern misused by changing an interface to fit this exact decorator.

Reply:
Maybe we should not think of it as a cache, but an authorization store?

Reply:
I think I have managed it. There is CachingOAuth2Service now, which is a decorator. However, it required a new service interface and a backward compatible wrapper.